File: Leetcode/Array&String/Array/11_container-with-most-water/main.py
from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:
        ans = 0
        l, r = 0, len(height) - 1
        while l < r:
            area = min(height[l], height[r]) * (r - l)
            ans = max(area, ans)
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        return ans


if __name__ == '__main__':
    heights = [1, 8, 6, 2, 5, 4, 8, 3, 7]
    res = Solution().maxArea(heights)
    print(res)

File: Leetcode/Array&String/Array/11_container-with-most-water/main.cpp
// 11. 盛最多水的容器
// Created by bob on 2022/4/21.
//
#include<vector>

using namespace std;

class Solution {
public:
    int maxArea(vector<int>& height) {
        int ans = 0;
        int l = 0, r = height.size() - 1;
        while (l < r) {
            int sum = min(height[l], height[r]) * (r - l);
            ans = max(sum, ans);
            if (height[l] <= height[r]) {
                ++l;
            } else {
                --r;
            }
        }
        return ans;
    }
};

int main() {
    vector<int> nums = {1, 8, 6, 2, 5, 4, 8, 3, 7};
    int res = Solution().maxArea(nums); // expect: 49
}